File: examples/chapter16/e02_int4_dot/python/main.py
"""Example 16.2 - The int4 Dot Product (NumPy reference + parity test).

Pack eight weights in -8..=7 as value+8 nibbles, two per byte, then take
their dot product with i16 activations two ways: subtracting 8 per weight,
and subtracting 8 * sum(activations) once. Same integer, as in rust.rs.

Run:   python main.py
Test:  pytest main.py
"""

import numpy as np

WEIGHTS = np.array([3, -2, 7, -8, 0, 5, -1, 4], dtype=np.int8)
ACTS = np.array([10, -3, 25, 7, -12, 4, 9, -6], dtype=np.int16)


def pack(w):
    n = (w.astype(np.int16) + 8).astype(np.uint8)
    return (n[0::2] | (n[1::2] << 4)).astype(np.uint8)


def build():
    codes = pack(WEIGHTS)
    nibbles = np.empty(8, dtype=np.int32)
    nibbles[0::2] = codes & 0xF
    nibbles[1::2] = codes >> 4
    a = ACTS.astype(np.int32)
    naive = int(((nibbles - 8) * a).sum())
    hoisted = int((nibbles * a).sum() - 8 * a.sum())
    real = np.float32(hoisted) * np.float32(0.05) * np.float32(0.02)
    return codes.tolist(), naive, hoisted, float(real)


def main():
    codes, naive, hoisted, real = build()
    print(f"8 weights -> {len(codes)} bytes:", " ".join(f"0x{b:02X}" for b in codes))
    print("naive dot   =", naive)
    print("hoisted dot =", hoisted)
    print("as a float  =", real)


def test_matches_burn():
    codes, naive, hoisted, real = build()
    assert codes == [0x6B, 0x0F, 0xD8, 0xC7]
    assert naive == 142
    assert hoisted == naive
    assert abs(real - 0.142) < 1e-6


if __name__ == "__main__":
    main()

File: examples/chapter16/e02_int4_dot/rust/src/main.rs
// Chapter 16 -- The int4 dot product at the heart of the firmware.
//
// esp32-tinyLLM stores every weight as a 4-bit code: value + 8, so -8..=7
// becomes 0..=15, and two codes share one byte (low nibble first). Its
// hottest loop, dot_int4_int16, multiplies those codes by i16 activations.
// This example is that idea at eight-element scale, with no framework, so
// every number can be checked by hand.

/// Pack weights in -8..=7 into bytes: two nibbles each, stored as value + 8.
fn pack(weights: &[i8]) -> Vec<u8> {
    weights
        .chunks(2)
        .map(|p| ((p[0] + 8) as u8) | (((p[1] + 8) as u8) << 4))
        .collect()
}

/// The obvious way: unpack each nibble, subtract 8, multiply, add.
/// One subtraction per weight.
fn dot_naive(codes: &[u8], acts: &[i16]) -> i32 {
    let mut acc = 0i32;
    for (i, &a) in acts.iter().enumerate() {
        let nibble = if i % 2 == 0 {
            codes[i / 2] & 0xF
        } else {
            codes[i / 2] >> 4
        };
        acc += (nibble as i32 - 8) * a as i32;
    }
    acc
}

/// The firmware's way: multiply the raw nibbles, and subtract 8 * sum(acts)
/// ONCE at the end. sum((n - 8) * a) = sum(n * a) - 8 * sum(a).
/// The activation sum is shared by every row, so it is computed once per token.
fn dot_hoisted(codes: &[u8], acts: &[i16], act_sum: i32) -> i32 {
    let mut acc = 0i32;
    for (p, &b) in codes.iter().enumerate() {
        acc += (b & 0xF) as i32 * acts[2 * p] as i32;
        acc += (b >> 4) as i32 * acts[2 * p + 1] as i32;
    }
    acc - 8 * act_sum
}

/// Returns (packed bytes, naive, hoisted, dequantised float result).
fn build() -> (Vec<u8>, i32, i32, f32) {
    let weights: [i8; 8] = [3, -2, 7, -8, 0, 5, -1, 4];
    let acts: [i16; 8] = [10, -3, 25, 7, -12, 4, 9, -6];
    let codes = pack(&weights);

    let act_sum: i32 = acts.iter().map(|&a| a as i32).sum();
    let naive = dot_naive(&codes, &acts);
    let hoisted = dot_hoisted(&codes, &acts, act_sum);

    // Back to real numbers: one scale for the weight group, one for the activations.
    let (w_scale, a_scale) = (0.05f32, 0.02f32);
    (codes, naive, hoisted, hoisted as f32 * w_scale * a_scale)
}

fn main() {
    let (codes, naive, hoisted, real) = build();
    let hex: Vec<String> = codes.iter().map(|b| format!("0x{b:02X}")).collect();
    println!("8 weights -> {} bytes: {}", codes.len(), hex.join(" "));
    println!("naive dot   = {naive}");
    println!("hoisted dot = {hoisted}");
    println!("as a float  = {real}");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn matches_numpy() {
        let (codes, naive, hoisted, real) = build();
        assert_eq!(codes, vec![0x6B, 0x0F, 0xD8, 0xC7]);
        assert_eq!(naive, 142);
        assert_eq!(hoisted, naive); // integer maths: exactly equal, not "close"
        assert!((real - 0.142).abs() < 1e-6);
    }
}
